framework/include/utils/OptionalMaterialPropertyProxy.h

```diff
@@ -0,0 +1,29 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "OptionalMaterialPropertyProxyForward.h"
+#include "Material.h"
+
+template <typename T>
+void
+OptionalMaterialPropertyProxy<T>::resolve()
+{
+  if (_material->hasMaterialProperty<T>(_name))
+    _value = &_material->getMaterialProperty<T>(_name);
```

Comment:
```suggestion
  if (_material->hasMaterialProperty<T>(_name))
  {
      mooseAssert(!_value, "Value already resolved");
      _value = &_material->getMaterialProperty<T>(_name);
   }
```

sanity check? maybe?

Reply:
That'd be a weak check. For one, resolving twice doesn't hurt at all, and the check would not detect already resolved but non-existing properties.

Reply:
Sounds good